File: DSA/DSA_450/Cpp/10_stack_and_queues/025_Circular_Queue.cpp
/*
Problem: Implement Circular Queue
URL: https://www.geeksforgeeks.org/circular-queue-set-1-introduction-array-implementation/

Problem Statement:
Implement circular queue with enqueue, dequeue, display using array with front/rear wrapping.

Sample Input/Output:
Input: enqueue(1), enqueue(2), enqueue(3), dequeue()
Output: 1
*/

#include <bits/stdc++.h>
using namespace std;

class CircularQueue {
private:
    int* arr;
    int front;
    int rear;
    int size;
    int capacity;

public:
    CircularQueue(int cap) : capacity(cap), front(-1), rear(-1), size(0) {
        arr = new int[capacity];
    }

    ~CircularQueue() {
        delete[] arr;
    }

    bool Is_Full() {
        return size == capacity;
    }

    bool Is_Empty() {
        return size == 0;
    }

    void Enqueue(int x) {
        if (Is_Full()) {
            cout << "Queue is full" << endl;
            return;
        }
        if (Is_Empty()) {
            front = rear = 0;
        } else {
            rear = (rear + 1) % capacity;
        }
        arr[rear] = x;
        size++;
    }

    int Dequeue() {
        if (Is_Empty()) {
            cout << "Queue is empty" << endl;
            return -1;
        }
        int val = arr[front];
        if (front == rear) {
            front = rear = -1;
        } else {
            front = (front + 1) % capacity;
        }
        size--;
        return val;
    }

    int Front_Element() {
        if (Is_Empty()) return -1;
        return arr[front];
    }

    int Rear_Element() {
        if (Is_Empty()) return -1;
        return arr[rear];
    }

    void Display() {
        if (Is_Empty()) {
            cout << "Queue is empty" << endl;
            return;
        }
        int i = front;
        cout << "Queue: ";
        while (true) {
            cout << arr[i] << " ";
            if (i == rear) break;
            i = (i + 1) % capacity;
        }
        cout << endl;
    }
};

class Solution {
public:
    void Test_Circular_Queue() {
        CircularQueue cq(5);
        
        cq.Enqueue(1);
        cq.Enqueue(2);
        cq.Enqueue(3);
        cq.Display();
        
        cout << "Dequeue: " << cq.Dequeue() << endl;
        cout << "Front: " << cq.Front_Element() << endl;
        cout << "Rear: " << cq.Rear_Element() << endl;
        
        cq.Enqueue(4);
        cq.Enqueue(5);
        cq.Enqueue(6);
        cq.Display();
        
        cout << "Dequeue: " << cq.Dequeue() << endl;
        cout << "Dequeue: " << cq.Dequeue() << endl;
        cq.Display();
    }
};

void Test_Circular_Queue() {
    Solution solution;
    solution.Test_Circular_Queue();
}

int main() {
    Test_Circular_Queue();
    return 0;
}

File: DSA/DSA_450/Python/10_stack_and_queues/025_Circular_Queue.py
"""
Problem: Implement Circular Queue
URL: https://www.geeksforgeeks.org/circular-queue-set-1-introduction-array-implementation/

Problem Statement:
Implement circular queue with enqueue, dequeue, display using array with front/rear wrapping.

Sample Input/Output:
Input: enqueue(1), enqueue(2), enqueue(3), dequeue()
Output: 1
"""


class CircularQueue:
    def __init__(self, cap):
        self.capacity = cap
        self.arr = [0] * self.capacity
        self.front = -1
        self.rear = -1
        self.size = 0

    def Is_Full(self):
        """
        Check if queue is full.
        Time Complexity: O(1)
        Space Complexity: O(1)
        """
        return self.size == self.capacity

    def Is_Empty(self):
        """
        Check if queue is empty.
        Time Complexity: O(1)
        Space Complexity: O(1)
        """
        return self.size == 0

    def Enqueue(self, x):
        """
        Add element to queue.
        Time Complexity: O(1)
        Space Complexity: O(1)
        """
        if self.Is_Full():
            print("Queue is full")
            return
        if self.Is_Empty():
            self.front = self.rear = 0
        else:
            self.rear = (self.rear + 1) % self.capacity
        self.arr[self.rear] = x
        self.size += 1

    def Dequeue(self):
        """
        Remove element from queue.
        Time Complexity: O(1)
        Space Complexity: O(1)
        """
        if self.Is_Empty():
            print("Queue is empty")
            return -1
        val = self.arr[self.front]
        if self.front == self.rear:
            self.front = self.rear = -1
        else:
            self.front = (self.front + 1) % self.capacity
        self.size -= 1
        return val

    def Front_Element(self):
        """
        Get front element without removing.
        Time Complexity: O(1)
        Space Complexity: O(1)
        """
        if self.Is_Empty():
            return -1
        return self.arr[self.front]

    def Rear_Element(self):
        """
        Get rear element without removing.
        Time Complexity: O(1)
        Space Complexity: O(1)
        """
        if self.Is_Empty():
            return -1
        return self.arr[self.rear]

    def Display(self):
        """
        Display queue elements.
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        if self.Is_Empty():
            print("Queue is empty")
            return
        i = self.front
        print("Queue: ", end="")
        while True:
            print(self.arr[i], end=" ")
            if i == self.rear:
                break
            i = (i + 1) % self.capacity
        print()


class Solution:
    def Test_Circular_Queue(self):
        cq = CircularQueue(5)
        
        cq.Enqueue(1)
        cq.Enqueue(2)
        cq.Enqueue(3)
        cq.Display()
        
        print(f"Dequeue: {cq.Dequeue()}")
        print(f"Front: {cq.Front_Element()}")
        print(f"Rear: {cq.Rear_Element()}")
        
        cq.Enqueue(4)
        cq.Enqueue(5)
        cq.Enqueue(6)
        cq.Display()
        
        print(f"Dequeue: {cq.Dequeue()}")
        print(f"Dequeue: {cq.Dequeue()}")
        cq.Display()


def Test_Circular_Queue():
    solution = Solution()
    solution.Test_Circular_Queue()


if __name__ == "__main__":
    Test_Circular_Queue()
